lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/Train.java

```diff
@@ -880,53 +880,57 @@ public void run() {
+                                String parametersPath = getName();
+                                if (Main.classDirectory != null)
+                                    parametersPath =
+                                            Main.classDirectory + File.separator + parametersPath;
+                                parametersPath += ".p";
+    
+                                Learner.Parameters bestParameters = tune();
+                                trainingRounds = bestParameters.rounds;
+                                Learner.writeParameters(bestParameters, parametersPath);
                                 System.out.println("  " + getName()
                                         + ": Training on entire training set");
+                            } else {
+                                if (lce.rounds != null)
+                                    trainingRounds = Integer.parseInt(((Constant) lce.rounds).value);
+    
+                                if (lce.K != null) {
+                                    int[] rounds = {trainingRounds};
+                                    int k = Integer.parseInt(lce.K.value);
+                                    double alpha = Double.parseDouble(lce.alpha.value);
+                                    trainer.crossValidation(rounds, k, lce.splitPolicy, alpha,
+                                            testingMetric, true);
+                                    System.out.println("  " + getName()
+                                            + ": Training on entire training set");
+                                }
                             }
-                        }
-
-                        trainer.train(lce.startingRound, trainingRounds);
-
-                        if (testParser != null) {
-                            System.out.println("Testing " + getName());
-                            new Accuracy(true).test(learner, learner.getLabeler(), testParser);
-                        }
-
-                        System.out.println("Writing " + getName());
-                    } else
-                        learner.saveLexicon(); // Writes .lex even if lexicon is empty.
-
+                            trainer.train(lce.startingRound, trainingRounds);
+                        } else
+                            learner.saveLexicon(); // Writes .lex even if lexicon is empty.
+                    } finally {
+                        learner.doneTraining();
+                    }
+                    
+                    if (lce.parser != null && testParser != null) {
+                        System.out.println("Testing " + getName());
+                        new Accuracy(true).test(learner, learner.getLabeler(), testParser);
+                    }
+                    
+                    // save the final model.
+                    System.out.println("Writing " + getName());
                     learner.save(); // Doesn't write .lex if lexicon is empty.
```

Comment:
@cowchipkid can you clarify the change here? Confused about what's happening here ....

Reply:
I moved the Accuracy reporting outside the block to ensure doneTraining() gets called before accuracy reporting. doneTraining will apply the feature optimization, we want the score after that. If accuracy is reported before doneTraining, we will get the score of the un-optimized models, which would be for not.

Reply:
Ah I see. Makes sense.